File: CFOnlineShop/CFOnlineShop/ViewController/CFBaseController/CFBaseController.m
//
//  CFBaseController.m
//  CFOnlineShop
//
//  Created by chenfeng on 2018/7/18.
//  Copyright © 2018年 chenfeng. All rights reserved.
//

#import "CFBaseController.h"
#import "CFTabBarController.h"

@interface CFBaseController ()

@end

@implementation CFBaseController

- (instancetype)init
{
    self = [super init];
    if (self) {
        
        self.hidesBottomBarWhenPushed = YES;
        
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.automaticallyAdjustsScrollViewInsets = NO;
    self.view.backgroundColor = KBackgroundColor;
    
    [self setNavView];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    //适配左滑时tabbar显示出来的问题
    if (self.hidesBottomBarWhenPushed) {
        if ([self.tabBarController isKindOfClass:[CFTabBarController class]]) {
            
            CFTabBarController *tabBar = (CFTabBarController *)self.tabBarController;
            [UIView animateWithDuration:animated ? 0.3 : 0 animations:^{
                tabBar.tabBarItemView.mj_y = Main_Screen_Height;
            }];
        }
    }
}

- (void)viewDidLayoutSubviews
{
    [self.view bringSubviewToFront:_navigationView];//始终放在最上层
}

- (void)setNavView
{
    _navigationView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, TopHeight)];
    _navigationView.backgroundColor = kClearColor;
    [self.view addSubview:_navigationView];
    
    _navigationBgView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, Main_Screen_Width, TopHeight)];
    _navigationBgView.backgroundColor = kClearColor;
    [_navigationView addSubview:_navigationBgView];
}

- (void)setTitle:(NSString *)title
{
    [_navigationView addSubview:self.titleLabel];
    _titleLabel.text = title;
}

- (UILabel *)titleLabel
{
    if (!_titleLabel) {
        
        _titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(_navigationView.mj_w/2 - 50, _navigationView.mj_h - 30, 100, 20)];
        _titleLabel.font = SYSTEMFONT(16);
        _titleLabel.textColor = KDarkTextColor;
        _titleLabel.textAlignment = NSTextAlignmentCenter;
    }
    
    return _titleLabel;
}

- (void)showLeftBackButton
{
    _leftButton = [UIButton buttonWithType:(UIButtonTypeCustom)];
    _leftButton.frame = CGRectMake(15, StatusBarHeight + 5, 45, 35);
    [_leftButton setImage:[UIImage imageNamed:@"back_btn"] forState:(UIControlStateNormal)];
    [_leftButton addTarget:self action:@selector(backAction) forControlEvents:(UIControlEventTouchUpInside)];
    [_navigationView addSubview:_leftButton];
}


#pragma mark - action

- (void)backAction
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftOnlineShopDemo/SwiftOnlineShopDemo/ViewController/CFBaseController/CFBaseController.swift
//
//  CFBaseController.swift
//  SwiftOnlineShopDemo
//
//  Created by chenfeng on 2019/4/23.
//  Copyright © 2019年 chenfeng. All rights reserved.
//

import UIKit

class CFBaseController: UIViewController {
    
    var navigationView: UIView!
    var navigationBgView: UIView!
    var leftButton: UIButton!
    var rightButton: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.white
     
        navigationView = UIView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: STATUS_AND_NAV_BAR_HEIGHT))
        navigationView.backgroundColor = UIColor.clear
        view.addSubview(navigationView)
        
        navigationBgView = UIView.init(frame: CGRect(x: 0, y: 0, width: SCREEN_WIDTH, height: STATUS_AND_NAV_BAR_HEIGHT))
        navigationBgView.backgroundColor = UIColor.clear
        navigationView.addSubview(navigationBgView)
        
        // Do any additional setup after loading the view.
    }
    
    override func viewDidLayoutSubviews() {
        self.view.bringSubviewToFront(navigationView)
    }
    
    func setTitle(title:String) {
        
        let titleLabel = UILabel.init(frame: CGRect(x: navigationView.mj_w/2 - 50, y: navigationView.mj_h - 30, width: 100, height: 20))
        titleLabel.font = UIFont.systemFont(ofSize: 16)
        titleLabel.textColor = UIColor.black
        titleLabel.text = title
        titleLabel.textAlignment = .center
        navigationView.addSubview(titleLabel)
        
    }

    func showLeftBackButton() {
        
        leftButton = UIButton(type:.custom)
        leftButton.frame = CGRect(x: 15, y: STATUSBAR_HEIGHT, width: 45, height: 35)
        leftButton.setImage(UIImage.init(named: "back_btn"), for:.normal)
        leftButton.addTarget(self, action: #selector(backAction), for: .touchUpInside)
        navigationView.addSubview(leftButton)
        
    }
    
    @objc func backAction() {
        self.navigationController?.popViewController(animated: true)
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
